src/Plugin/Commerce/PaymentGateway/ExpressCheckout.php

```diff
@@ -0,0 +1,570 @@
+    ];
+
+    $form['solution_type'] = [
+      '#type' => 'radios',
+      '#title' => t('Type of checkout flow'),
+      '#description' => t('Express Checkout Account Optional (ECAO) where PayPal accounts are not required for payment may not be available in all markets.'),
+      '#options' => [
+        'Mark' => t('Require a PayPal account (this is the standard configuration).'),
+        'SoleLogin' => t('Allow PayPal AND credit card payments, defaulting to the PayPal form.'),
+        'SoleBilling' => t('Allow PayPal AND credit card payments, defaulting to the credit card form.'),
+      ],
+      '#default_value' => $this->configuration['solution_type'],
+    ];
+
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitConfigurationForm(array &$form, FormStateInterface $form_state) {
+    parent::submitConfigurationForm($form, $form_state);
+    if (!$form_state->getErrors()) {
+      $values = $form_state->getValue($form['#parents']);
+      $this->configuration['api_username'] = $values['api_username'];
+      $this->configuration['api_password'] = $values['api_password'];
+      $this->configuration['signature'] = $values['signature'];
+      $this->configuration['solution_type'] = $values['solution_type'];
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function onReturn(OrderInterface $order, Request $request) {
+    // GetExpressCheckoutDetails API Operation (NVP).
+    // Shows information about an Express Checkout transaction.
+    $paypal_response = $this->apiGetExpressCheckoutDetails($order);
+
+    // If the request failed, exit now with a failure message.
+    if ($paypal_response['ACK'] == 'Failure') {
+      return FALSE;
+    }
+
+    // Set the Payer ID used to finalize payment.
+    $order_express_checkout_data = $order->getData('paypal_express_checkout');
+    $order_express_checkout_data['payerid'] = $paypal_response['PAYERID'];
+    $order->setData('paypal_express_checkout', $order_express_checkout_data);
+
+    // If the user is anonymous, add their PayPal e-mail to the order.
+    if (empty($order->mail)) {
+      $order->setEmail($paypal_response['EMAIL']);
+    }
+    $order->save();
+
+    // DoExpressCheckoutPayment API Operation (NVP).
+    // Completes an Express Checkout transaction.
+    $paypal_response = $this->apiDoExpressCheckoutDetails($order);
+
+    // Nothing to do for failures for now - no payment saved.
+    // ToDo - more about the failures.
+    if ($paypal_response['PAYMENTINFO_0_PAYMENTSTATUS'] == 'Failed') {
+      return FALSE;
+    }
+
+    $payment_storage = $this->entityTypeManager->getStorage('commerce_payment');
+    $payment = $payment_storage->create([
+      'state' => 'authorization',
+      'amount' => $order->getTotalPrice(),
+      'payment_gateway' => $this->entityId,
+      'order_id' => $order->id(),
+      'test' => $this->getMode() == 'test',
+      'remote_id' => $paypal_response['PAYMENTINFO_0_TRANSACTIONID'],
```

Comment:
When reference transactions are enabled:
Since reference transactions are using MerchantInitiatedBillingSingleAgreement, the BILLINGAGREEMENTID coming back should be stored (as a commerce_payment_method?) so that it can be used with DoReferenceTransaction to do recurring billing, click-to-buy, etc. in the future, without the 730 day limitation that using the transaction id as the reference imposes.